src/libos/crates/async-rt/src/time/timer.rs

```diff
@@ -0,0 +1,200 @@
+pub use core::task::Waker as RawWaker;
+
+use crate::prelude::*;
+use crate::time::Instant;
+
+use hierarchical_hash_wheel_timer::wheels::quad_wheel::QuadWheelWithOverflow;
+
+lazy_static! {
+    pub static ref TIMER_WHEEL: TimerWheel = TimerWheel::new();
+}
+
+/// The interface of hierarchical timer wheel.
+/// The resolution of time is 1ms.
+pub struct TimerWheel {
+    // hierarchical timer wheel.
+    wheel: Mutex<QuadWheelWithOverflow<TimerWheelEntry>>,
+    // current tick, one tick means 1ms.
+    ticks: AtomicU64,
+    // start time of the timerwheel.
+    start: Instant,
+}
+
+impl TimerWheel {
+    pub fn new() -> Self {
+        Self {
+            wheel: Mutex::new(QuadWheelWithOverflow::default()),
+            ticks: AtomicU64::new(0),
+            start: Instant::now(),
+        }
+    }
+
+    /// Try to move the timerwheel forward.
+    /// If get the wheel lock successfully, update the ticks and the wheel.
+    pub fn try_make_process(&self) {
+        match self.wheel.try_lock() {
+            Some(mut guard) => self.make_process(&mut guard),
+            None => {
+                return;
+            }
+        }
+    }
+
+    /// Insert a new timer entry to the wheel.
+    /// Return the latest tick of the timerwheel, or Err() if the duration is too short (less than 1ms).
+    pub fn insert_entry(&self, entry: TimerWheelEntry, timeout: Duration) -> Result<u64> {
+        let mut guard = self.wheel.lock();
+        // make process first, since the try_make_process() may often not get the lock.
+        self.make_process(&mut guard);
+
+        match guard.insert_with_delay(entry, timeout) {
+            Ok(_) => Ok(self.latest_tick()),
+            Err(_) => Err("wheel.insert_with_delay() error."),
+        }
+    }
+
+    /// Move the timerwheel forward.
+    fn make_process(&self, wheel_guard: &mut MutexGuard<QuadWheelWithOverflow<TimerWheelEntry>>) {
+        // We use the untrusted clock. If we get non-decreasing time unfortunately, use 0 instead.
+        let elapsed = self
+            .start
+            .checked_elapsed()
+            .map_or(0, |e| e.as_millis() as u64);
```

Comment:
If the clock is decreasing, which is an unexpected and non-recoverable error, I think we should simply panic, instead of hiding  such abnormal circumstances.

Reply:
If we use `LAST_NOW`, the clock will be non-decreasing.